### DISCUSSION TITLE
If --compile-commands-dir is used, should clangd load the index on startup?

### DISCUSSION THREAD

**Discussion by HighCommander4** [clangd/clangd#1341] [General]
Currently, clangd loads the index when first opening a file.

IIUC, the reason for this is that the index is stored next to the `compile_commands.json`, and the location of the `compile_commands.json` is (usually) determined by the path of the opened file.

However, in the case where clangd is run with `--compile-commands-dir`, the location of the `compile_commands.json` which will be used for all open files is already known on startup. In this case, could clangd load the index on startup (or after the `initialize` request, but without waiting for the first file to be opened)?

**Comment by Yurii-huang** [comment]
> Currently, clangd loads the index when first opening a file.
> 
> IIUC, the reason for this is that the index is stored next to the `compile_commands.json`, and the location of the `compile_commands.json` is (usually) determined by the path of the opened file.
> 
> However, in the case where clangd is run with `--compile-commands-dir`, the location of the `compile_commands.json` which will be used for all open files is already known on startup. In this case, could clangd load the index on startup (or after the `initialize` request, but without waiting for the first file to be opened)?

It seems that Java does this. It will start some important data loading and simultaneously synchronize the front and rear ends with the LSP maintaining percentage progress communication

**Reply by Yurii-huang** [reply]
java lsp backend is jtdls

**Comment by HighCommander4** [comment]
It's worth noting here that the usefulness of this proposed behaviour is entangled with the client's behaviour in when it starts the server. Using VSCode as an example, it currently only starts the server when a C/C++ file is opened (or if a command to start the server is explicitly invoked by the user).

If the client only starts the server when a C/C++ file is opened, we don't get much out of clangd loading the index on startup rather than when a C/C++ file is opened (since those two things will happen together anyways).

I think a good way forward could be to have the "strong workspace mode" discussed in https://github.com/clangd/clangd/discussions/907 change both behaviours (when the client starts the server, and when the server loads the index). (This in turn implies that "strong workspace mode" needs to be a client-side configuration, which it can propagate to the server via `initialize`, rather than a server-side configuration. That's probably better anyways, as strong workspace mode may not play well with clients that don't expect it.)
